File: python/valid_parenthesis_string.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        leftMin: int = 0
        leftMax: int = 0
        for c in s:
            if c == "(":
                leftMin += 1
                leftMax += 1
            elif c == ")":
                leftMin -= 1
                leftMax -= 1
            else:
                leftMin -= 1
                leftMax += 1

            if leftMax < 0:
                return False
            if leftMin < 0:
                leftMin = 0

        return leftMin == 0

    def checkValidString2(self, s: str) -> bool:
        """time complexity: O(n), space complexity: O(n)"""
        left_stack: list[int] = []
        star_stack: list[int] = []
        for i, c in enumerate(s):
            if c == "(":
                left_stack.append(i)
            elif c == "*":
                star_stack.append(i)
            else:
                if len(left_stack) == 0 and len(star_stack) == 0:
                    return False
                if len(left_stack) > 0:
                    left_stack.pop()
                else:
                    star_stack.pop()
        while len(left_stack) > 0 and len(star_stack) > 0:
            if left_stack.pop() > star_stack.pop():
                return False
        return len(left_stack) == 0


def test_checkValidString_case_1():
    # arrange
    s = "()"
    expected = True

    # act
    actual = Solution().checkValidString(s)

    # assert
    assert actual == expected


def test_checkValidString_case_2():
    # arrange
    s = "(*)"
    expected = True

    # act
    actual = Solution().checkValidString(s)

    # assert
    assert actual == expected


def test_checkValidString_case_3():
    # arrange
    s = "(*))"
    expected = True

    # act
    actual = Solution().checkValidString(s)

    # assert
    assert actual == expected

File: rust/src/code/valid_parenthesis_string.rs
struct Solution {}

impl Solution {
    pub fn check_valid_string(s: String) -> bool {
        let s: Vec<char> = s.chars().collect();
        let mut left_min = 0;
        let mut left_max = 0;

        for c in s {
            if c == '(' {
                left_max += 1;
                left_min += 1;
            } else if c == ')' {
                left_max -= 1;
                left_min -= 1;
            } else if c == '*' {
                left_max += 1;
                left_min -= 1;
            }

            if left_max < 0 {
                return false;
            }

            if left_min < 0 {
                left_min = 0;
            }
        }

        return left_min == 0;
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn test_chech_valid_string_case_1() {
        // arrange
        let s = String::from("()");
        let expected = true;

        // act
        let actual = Solution::check_valid_string(s);

        // assert
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_chech_valid_string_case_2() {
        // arrange
        let s = String::from("(*)");
        let expected = true;

        // act
        let actual = Solution::check_valid_string(s);

        // assert
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_chech_valid_string_case_3() {
        // arrange
        let s = String::from("(*))");
        let expected = true;

        // act
        let actual = Solution::check_valid_string(s);

        // assert
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_chech_valid_string_case_4() {
        // arrange
        let s = String::from("(((((*(()((((*((**(((()()*)()()()*((((**)())*)*)))))))(())(()))())((*()()(((()((()*(())*(()**)()(())");
        let expected = false;

        // act
        let actual = Solution::check_valid_string(s);

        // assert
        assert_eq!(actual, expected);
    }
}
